From 1392fde1ade2c9caf60bd1bf65c684d81749ed15 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Wed, 13 Jun 2018 18:22:58 +0800
Subject: [PATCH 4255/5242] MLK-18605-2 drm/bridge: sec-dsim: allow same panel
 can be re-attached

commit  0a7a510229216d252c3c3b354ee1a40549588973 from
https://source.codeaurora.org/external/imx/linux-imx.git

During the mode set procedure, some dsi client device may detach
itself first and then attach it again according to the target
display mode parameters. In this case, the dsi client device
should be allowed to be re-attached again. So this is also true
for panel device.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit f586625dd1a665c58b976405c7980b7414554481)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/bridge/sec-dsim.c |   12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/bridge/sec-dsim.c b/drivers/gpu/drm/bridge/sec-dsim.c
index 5e89a67..b2f0ade 100644
--- a/drivers/gpu/drm/bridge/sec-dsim.c
+++ b/drivers/gpu/drm/bridge/sec-dsim.c
@@ -272,10 +272,6 @@ static int sec_mipi_dsim_host_attach(struct mipi_dsi_host *host,
 	struct device *dev = dsim->dev;
 	struct drm_panel *panel;
 
-	/* Don't support multiple panels */
-	if (dsim->panel)
-		return -EBUSY;
-
 	if (!dsi->lanes || dsi->lanes > pdata->max_data_lanes) {
 		dev_err(dev, "invalid data lanes number\n");
 		return -EINVAL;
@@ -299,11 +295,17 @@ static int sec_mipi_dsim_host_attach(struct mipi_dsi_host *host,
 		return -EINVAL;
 	}
 
-	if (!dsim->next && !dsim->panel) {
+	if (!dsim->next) {
 		/* 'dsi' must be panel device */
 		panel = of_drm_find_panel(dsi->dev.of_node);
 		WARN_ON(!panel);
 
+		/* Don't support multiple panels */
+		if (dsim->panel && panel && dsim->panel != panel) {
+			dev_err(dev, "don't support multiple panels\n");
+			return -EBUSY;
+		}
+
 		dsim->panel = panel;
 	}
 
-- 
1.7.9.5

